<a href="https://colab.research.google.com/github/Jolanta27/numpy_exercises/blob/master/Time_to_Practice__Numpy_1_with_partial_solutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img height="180px" src="https://drive.google.com/uc?export=view&id=141XOz6N4nk8Ru1sAl7vOsAToCLrSFCAX" alt="SDA logo" align="left" hspace="30px" vspace="50px"/>

# Welcome to your next notebook with SDA!

During the classes we will mostly use [Google Colaboratory](https://colab.research.google.com/?hl=en) which is a free Jupyter notebook environment that requires no setup and runs entirely in the cloud.

However, for bigger projects, especially involving Deep Learning and/or big data reading, it might be a better choice to setup Jupyter Notebook or Jupyter Lab on your computer. Also, it is worth noticing that there is a great number of useful extensions (see [nbextensions](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/index.html) and [jupyter-labextension](https://jupyterlab.readthedocs.io/en/stable/user/extensions.html)) not available for Colab users.

<img src="https://drive.google.com/uc?export=view&id=1UO2urRciECzoKE_vHy4RMGfFbkOWOGlW" alt="SDA logo" align="left" width="100px" hspace="10px" vspace="10px"/>
<br>

# TASKS

<img src="https://drive.google.com/uc?export=view&id=141XOz6N4nk8Ru1sAl7vOsAToCLrSFCAX" alt="SDA logo" width="150" align='right'/>
<br>

## Linear Algebra

### **<font color='#306998'>TASKS </font><font color='#ffd33b'>1.1 - 1.10</font>** should be already completed :)

### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.11</font>**

For the matrix from task 1.7:
* calculate its determinant, (optional)
* calculate its trace (the sum of elements on the main diagonal),
* find the largest and smallest item.

$$
A=
\begin{bmatrix}
1 & 2 & 3 & 4 \\
5 & 6 & 7 & 8 \\
9 & 10 & 11 & 12 \\
13 & 14 & 15 & 16
\end{bmatrix}
$$

In [3]:
import numpy as np

In [4]:
A = np.arange(1, 17).reshape(4, 4)
A

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [5]:
A.trace(), A.diagonal().sum()

(34, 34)

In [6]:
A.max(), A.min()

(16, 1)

### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.12</font>**

Define two `3x3` real matrices, ie. whose elements consist entirely of real numbers. For the first matrix, draw the numbers from standard normal (Gaussian) distribution, ie. $\mu = 0$ and $\sigma = 1$. For the second one, set the standard devation to 1 and mean value to 5.

Next, for each of them:
* multiply all elements by a scalar 10,
* extract the element in the middle,
* transpose them
* get its trace & sum of all elements

and calculate their:
* sum
* difference
* element-wise multiplication result
* element-wise division result
* matrix product (matrix multiplication)

In [22]:
A = np.random.normal(size=9).reshape(3, 3)
print(A)
A = A.reshape(3, -1)
print(A)
A * 10
print(A)
B = A[1,1]
print(B)
print(A)
C = A.T
print(C)
D = A.trace()
print(D)

[[ 1.24308302 -0.94673477  1.28538196]
 [ 0.10795782 -0.51829729 -1.16129897]
 [ 0.14848363  0.55950117 -0.17466449]]
[[ 1.24308302 -0.94673477  1.28538196]
 [ 0.10795782 -0.51829729 -1.16129897]
 [ 0.14848363  0.55950117 -0.17466449]]
[[ 1.24308302 -0.94673477  1.28538196]
 [ 0.10795782 -0.51829729 -1.16129897]
 [ 0.14848363  0.55950117 -0.17466449]]
-0.518297292603109
[[ 1.24308302 -0.94673477  1.28538196]
 [ 0.10795782 -0.51829729 -1.16129897]
 [ 0.14848363  0.55950117 -0.17466449]]
[[ 1.24308302  0.10795782  0.14848363]
 [-0.94673477 -0.51829729  0.55950117]
 [ 1.28538196 -1.16129897 -0.17466449]]
0.5501212363569972


In [8]:
A = np.random.normal(size=9)
A

array([-1.33933051, -0.76244196, -0.93163352, -0.69863876,  1.45119433,
        1.84378929, -1.93789777, -0.60000051,  0.83006185])

In [10]:
B = np.random.normal(loc=5, scale=1, size=9).reshape(3, 3)
B

array([[4.76877661, 3.52198689, 4.61132691],
       [4.04505086, 6.24397212, 5.27425961],
       [5.98843567, 4.1704498 , 3.22828954]])

In [11]:
A * 10

array([[ 13.31719301,   1.97054974,   0.45174793],
       [ -6.50659387,  15.22028456, -11.00112056],
       [ -3.87119024,  -7.65421183,  -1.45804753]])

In [ ]:
A[1, 1]

-0.6504013241436976

In [24]:
A.T

array([[ 1.24308302,  0.10795782,  0.14848363],
       [-0.94673477, -0.51829729,  0.55950117],
       [ 1.28538196, -1.16129897, -0.17466449]])

In [23]:
A.trace()

0.5501212363569972

In [33]:
A.min()

-1.1612989733680907

In [34]:
A.max()

1.2853819648195486

In [37]:
B = np.random.normal(loc=5, scale=1, size=(3,3))
print(B)

[[5.21645568 4.50934587 4.41099422]
 [5.6294908  5.50402292 5.48931687]
 [7.21029692 6.48292646 5.19089301]]


In [38]:
A@B

array([[ 10.4228384 ,   8.72767815,   6.95858511],
       [-10.72790307,  -9.89451688,  -8.39707548],
       [  2.66488215,   2.61673428,   2.81957499]])

In [27]:
A.sum()

0.5434120774206814

In [28]:
A + B

array([[ 0.72478573, -1.46503206,  0.76708467],
       [-0.41033947, -1.03659459, -1.67959627],
       [-0.36981367,  0.04120388, -0.69296178]])

In [29]:
A - B

array([[ 1.76138031, -0.42843748,  1.80367926],
       [ 0.62625511,  0.        , -0.64300168],
       [ 0.66678092,  1.07779847,  0.3436328 ]])

In [30]:
A * B

array([[-0.64428656,  0.49069007, -0.66620999],
       [-0.05595425,  0.26863208,  0.60189811],
       [-0.07695866, -0.28998794,  0.09052813]])

In [31]:
A / B

array([[-2.39839767,  1.82662496, -2.48000903],
       [-0.20829323,  1.        ,  2.24060397],
       [-0.28648351, -1.07949855,  0.33699672]])

In [39]:
np.matmul(A, B)

array([[ 10.4228384 ,   8.72767815,   6.95858511],
       [-10.72790307,  -9.89451688,  -8.39707548],
       [  2.66488215,   2.61673428,   2.81957499]])

### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.14</font>**

Create an array with the following elements: `[23, 45, 124, 112, 150, 43, 25, 17, 268, 254, 95, 97, 8]` and filter out the ones that are less than or equal to their mean value. How many such numbers are there?

In [ ]:
array = np.array([23, 45, 124, 112, 150, 43, 25, 17, 268, 254, 95, 97, 8])
array

array([ 23,  45, 124, 112, 150,  43,  25,  17, 268, 254,  95,  97,   8])

In [ ]:
array[array <= array.mean()]

array([23, 45, 43, 25, 17, 95, 97,  8])

In [ ]:
len(array[array <= array.mean()])

8

### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.15</font>**

Use the array defined above to evaluate mean value and standard deviation. Check how many of these numbers is within a distance of at most one standard deviation from the mean, ie. belongs to the closed interval `[mean - std, mean + std]`.

In [ ]:
lower_limit = array.mean() - array.std()
upper_limit = array.mean() + array.std()

np.sum((array > lower_limit) & (array < upper_limit))

# to see which values belong to the interval
# array[(array > lower_limit) & (array < upper_limit)]
# len(...)

10

In [52]:
array = np.array([23, 45, 43, 25, 17, 95, 97, 8])
print(array)
mean = np.mean(array)
print(mean)
std = np.std(array)
print(std)
within_one_std = np.sum((array >= mean - std) & (array <= mean + std))
print(within_one_std)

[23 45 43 25 17 95 97  8]
44.125
32.091422140503525
5


In [ ]:
array[array <= array.mean()]

array([23, 45, 43, 25, 17, 95, 97,  8])

### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.16</font>**

Write a function `get_sum_of_squared_naturals()` that takes one argument as an input (call it `max_value`) and returns the sum of squared natural numbers `k` such that $k \leq max\_value.$

Make sure that the input given by a user is a natural number. Set the default value to 5.

In [53]:
def get_sum_of_squared_naturals(max_value=5):
    assert isinstance(max_value, int), "Input is not an integer"
    assert max_value >= 0, "Input is not a positive number"

    return np.sum(np.arange(max_value + 1) ** 2)


get_sum_of_squared_naturals()
# get_sum_of_squared_naturals(3)
# get_sum_of_squared_naturals(5.4)
# get_sum_of_squared_naturals(-1)

55

In [71]:
def get_sum_of_squared_naturals(max_value=5):
  if not isinstance(max_value, int) or max_value < 1:
    raise ValueError("max_value must be a natural number")

    naturals = np.arange(1, max_value + 1)
    sum_of_squares = np.sum(naturals**2)
    return sum_of_squares
print(get_sum_of_squared_naturals(max_value=5))

None


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.17</font>**

Create the following 4x4 matrix 4x4, and then:
* print out an element from the second row and the third column (starting from 1),
* check which element is equal to 0,
* with array indexing:
    * display element equal to 3,
    * view vector `[3, 11, 0]`,
* calculate its:
    * determinant,
    * trace (sum of diagonal entries),
    * the maximum and mininum elements,
    * the mean value of each column,
    * the maximum value of each row,
* find its transposition and multiply both matrices mathematically, ie. get the product of $A \cdot A^T,$
* determine its eigenvalues and eigenvectors. (optional)

```
np.array([
    [1, 15, 4, 13],
    [8, 21, 3, 12],
    [11, 13, 11, 5],
    [32, 13, 0, 2]
])
```

In [ ]:
A = np.array([
    [1, 15, 4, 13],
    [8, 21, 3, 12],
    [11, 13, 11, 5],
    [32, 13, 0, 2]
])

In [ ]:
A == 0

array([[False, False, False, False],
       [False, False, False, False],
       [False, False, False, False],
       [False, False,  True, False]])

In [ ]:
np.where(A == 11)

(array([2, 2]), array([0, 2]))

In [ ]:
A[1, 2]

3

In [ ]:
np.linalg.det(A)

-19963.99999999999

In [ ]:
A.mean(axis=0)

array([13. , 15.5,  4.5,  8. ])

In [ ]:
np.matmul(A, A.T)

array([[ 411,  491,  315,  253],
       [ 491,  658,  454,  553],
       [ 315,  454,  436,  531],
       [ 253,  553,  531, 1197]])

In [100]:
import numpy as np
from numpy.linalg import det, eig

A = np.array([[0,2,3,4], [5,6,7,8], [9,10,11,12], [13,14,15,16]])
print(A)
print("an element from the second row and the third column: ", A[1,2])
where_0 = np.where(A == 0)
print(where_0)
index = np.where(A == 3)
print(index)
vector = A[[0,2,2], [2,0,2]]
print("Vector [3,11,0] :", vector)
print("Determinant: ", det(A))
print("Trace: ", np.trace(A))
print(np.max(A))
print(np.min(A))
print("Mean value of each column: ", np.mean(A, axis=0))
print("Maximum value of each row: ", np.max(A, axis=1))
AT = A.T
print("Product of A and AT:\n ", np.dot(A, AT))
eigenvalues, eigenvectors = eig(A)
print("Eigenvalues: ", eigenvalues)
print("Eigenvectors: ", eigenvectors)


[[ 0  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]
an element from the second row and the third column:  7
(array([0]), array([0]))
(array([0]), array([2]))
Vector [3,11,0] : [ 3  9 11]
Determinant:  0.0
Trace:  33
16
0
Mean value of each column:  [ 6.75  8.    9.   10.  ]
Maximum value of each row:  [ 4  8 12 16]
Product of A and AT:
  [[ 29  65 101 137]
 [ 65 174 278 382]
 [101 278 446 614]
 [137 382 614 846]]
Eigenvalues:  [ 3.61446939e+01 -2.91706886e+00 -2.27625027e-01 -2.82998134e-16]
Eigenvectors:  [[-1.47329223e-01 -8.04983593e-01  4.08687234e-01  6.07085845e-16]
 [-3.49531408e-01 -1.73555370e-01 -7.67365771e-01  4.08248290e-01]
 [-5.47657498e-01  1.81916530e-01 -1.47978089e-01 -8.16496581e-01]
 [-7.45783588e-01  5.37388430e-01  4.71409592e-01  4.08248290e-01]]


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.18</font>**

Create a 3x3 array and normalize its columns, ie. scale the values such that sum of each column equals 1 (unit vectors wrt. Manhattan metric).

In [101]:
A = np.arange(7, 16).reshape(3, -1)
A

array([[ 7,  8,  9],
       [10, 11, 12],
       [13, 14, 15]])

In [102]:
A / A.sum(axis=0)

array([[0.23333333, 0.24242424, 0.25      ],
       [0.33333333, 0.33333333, 0.33333333],
       [0.43333333, 0.42424242, 0.41666667]])

In [103]:
(A / A.sum(axis=0)).sum(axis=0)

array([1., 1., 1.])

In [109]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(A)
A_normalized = A / A.sum(axis=0)
print(A_normalized)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[0.08333333 0.13333333 0.16666667]
 [0.33333333 0.33333333 0.33333333]
 [0.58333333 0.53333333 0.5       ]]


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.19</font>**

Write a function that accepts a positive natural number $p\in\mathbb{N}_1$ and two NumPy vectors as an input (both of the same shape). The function should return the Mikowski distance between those vectors, ie.:
$$d_p(X, Y) = \left(\sum_{i=1}^n |x_i - y_i|^p\right)^\frac{1}{p},$$

where $X=[x_1, x_2, ..., x_n]\in\mathbb{R}^n$ and $Y=[y_1, y_2, ..., y_n]\in\mathbb{R}^n.$

*Hint: you can use the following assertions `assert a.size == b.size, "Arrays must be of the same shape"`, and `assert isinstance(input_array_1, np.ndarray)`.*

In [116]:
X = np.array([1,2,3])
Y = np.array([4,5,6])

def minkowski_distance(p, X, Y):
  assert isinstance(X, np.ndarray), "Input must be a numpy array"
  assert isinstance(Y, np.ndarray), "Input must be a numpy array"
  assert X.size == Y.size, "Arrays must be of the same shape"
  assert p > 0, "p must be a positive natural number"

  return np.sum(np.abs(X - Y) ** p) ** (1 / p)
print(minkowski_distance(2, X, Y))

5.196152422706632


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.20</font>**

Determine the length of the following vectors using Euclidean norm:
* $v_1=[0, 5]$
* $v_2=[0, -3]$
* $v_3=[7, 0]$
* $v_4=[-\pi, 0]$
* $v_5=[3, -4]$
* $v_6=[1, 7]$
* $v_7=[5\sqrt{3}, -3, 4]$
* $v_8=[1, 2, 3, 4]$

*Hint: using default parameters of `numpy.linalg.norm()` function will return Frobenius norm for matrices which is equivalent to Euclidean norm for vectors:*
https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html

In [ ]:
v1 = np.array([0, 5])
v1_len = np.linalg.norm(v1)

v1, v1_len

(array([0, 5]), 5.0)

In [119]:
v1 = np.array([0, 5])
v2 = np.array([0, -3])
v3 = np.array([7, 0])
v4 = np.array([-np.pi, 0])
v5 = np.array([3, -4])
v6 = np.array([1, 7])
v7 = np.array([np.sqrt(53), -3, 4])
v8 = np.array([1,2,3,4])
print("Euclidean norm of v1: ", np.linalg.norm(v1))
print("Euclidean norm of v1: ", np.linalg.norm(v2))
print("Euclidean norm of v1: ", np.linalg.norm(v3))
print("Euclidean norm of v1: ", np.linalg.norm(v4))
print("Euclidean norm of v1: ", np.linalg.norm(v5))
print("Euclidean norm of v1: ", np.linalg.norm(v6))
print("Euclidean norm of v1: ", np.linalg.norm(v7))
print("Euclidean norm of v1: ", np.linalg.norm(v8))


Euclidean norm of v1:  5.0
Euclidean norm of v1:  3.0
Euclidean norm of v1:  7.0
Euclidean norm of v1:  3.141592653589793
Euclidean norm of v1:  5.0
Euclidean norm of v1:  7.0710678118654755
Euclidean norm of v1:  8.831760866327848
Euclidean norm of v1:  5.477225575051661


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.21</font>**

For the following vectors `a=[3, -2]` and `b=[-2, -3]`, determine their scalar product and the angle between these vectors.

*Correct answer $\alpha = 90^\circ$.*

In [120]:
a = np.array([3, -2])
b = np.array([-2, -3])

np.dot(a, b)

0

In [121]:
denominator = np.linalg.norm(a) * np.linalg.norm(b)
denominator

12.999999999999998

In [122]:
cos_gamma = np.dot(a, b) / denominator
cos_gamma

0.0

In [123]:
gamma = np.arccos(cos_gamma)
gamma  # radians, ~pi/2

1.5707963267948966

In [124]:
a = np.array([3, -2])
b = np.array([-2, -3])
scalar_product = np.dot(a, b)
norm_a = np.linalg.norm(a)
norm_b = np.linalg.norm(b)
angle_rad = np.arccos(scalar_product / (norm_a * norm_b))
angle_deg = np.degrees(angle_rad)
print("Scalar_product: ", scalar_product)
print("Angle (degrees): ", angle_deg)

Scalar_product:  0
Angle (degrees):  90.0


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.22</font>**

For the following vectors `a=[0, 3]` and `b=[2, 2]`, determine their scalar product and the angle between these vectors.

*Correct answer $\alpha = 45^\circ$.*

In [125]:
a = np.array([0, 3])
b = np.array([2, 2])
scalar_product = np.dot(a,b)
norm_a = np.linalg.norm(a)
norm_b = np.linalg.norm(b)
angle_rad = np.arccos(scalar_product / (norm_a * norm_b))
angle_deg = np.degrees(angle_rad)
print("scalar_product: ", scalar_product)
print("Angle (degrees): ", angle_deg)


scalar_product:  6
Angle (degrees):  45.00000000000001


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.23</font>**

For the following vectors `a=[3, -2]` and `b=[1, 7]`, determine their scalar product and the angle between these vectors.

*Correct answer $\alpha ≈ 116^\circ$.*

In [127]:
a = np.array([3, -2])
b = np.array([1, 7])
scalar_product = np.dot(a,b)
norm_a = np.linalg.norm(a)
norm_b = np.linalg.norm(b)
angle_rad = np.arccos(scalar_product / (norm_a * norm_b))
angle_deg = np.degrees(angle_rad)
print("scalar_product: ", scalar_product)
print("Angle (degrees): ", angle_deg)

scalar_product:  -11
Angle (degrees):  115.55996517182382


### **<font color='#306998'>TASK </font><font color='#ffd33b'>1.24</font>**

Normalize each vector from Task 20.

In [ ]:
v1 / v1_len

array([0., 1.])

In [130]:
v_1 = np.array([0,5])
v_2 = np.array([0,-3])
v_3 = np.array([7,0])
v_4 = np.array([-np.pi,0])
v_5 = np.array([3,-4])
v_6 = np.array([1,7])
v_7 = np.array([np.sqrt(53),-3,4])
v_8 = np.array([1,2,3,4])

v1_normalized = v1 / np.linalg.norm(v1)
v2_normalized = v2 / np.linalg.norm(v2)
v3_normalized = v3 / np.linalg.norm(v3)
v4_normalized = v4 / np.linalg.norm(v4)
v5_normalized = v5 / np.linalg.norm(v5)
v6_normalized = v6 / np.linalg.norm(v6)
v7_normalized = v7 / np.linalg.norm(v7)
v8_normalized = v8 / np.linalg.norm(v8)
print("V1 normalized: ", v1_normalized)
print("V2 normalized: ", v2_normalized)
print("V3 normalized: ", v3_normalized)
print("V4 normalized: ", v4_normalized)
print("V5 normalized: ", v5_normalized)
print("V6 normalized: ", v6_normalized)
print("V7 normalized: ", v7_normalized)
print("V8 normalized: ", v8_normalized)

V1 normalized:  [0. 1.]
V2 normalized:  [ 0. -1.]
V3 normalized:  [1. 0.]
V4 normalized:  [-1.  0.]
V5 normalized:  [ 0.6 -0.8]
V6 normalized:  [0.14142136 0.98994949]
V7 normalized:  [ 0.82431012 -0.33968311  0.45291081]
V8 normalized:  [0.18257419 0.36514837 0.54772256 0.73029674]


<img src="https://drive.google.com/uc?export=view&id=141XOz6N4nk8Ru1sAl7vOsAToCLrSFCAX" alt="SDA logo" width="150" align='right'/>
<br>

## Appendix: random numbers generators

1. Randomly generated number(s) from normal (Gauss) distribution.
```python
np.random.normal()
np.random.normal(size=6)
```
1. Randomly selected numbers from the uniform system.
```python
np.random.uniform(size=6)
```
1. Randomly selected number from a specific range.
```python
np.random.randint(3, 30, size=6)
```
1. Randomly selected floating-point number
```python
np.random.random()
```